lib/t/feeds.ex

```diff
@@ -97,42 +97,110 @@ defmodule T.Feeds do
         ) ::
           {[feed_profile], feed_cursor}
   def fetch_feed(user_id, location, gender, gender_preferences, count, feed_cursor) do
-    feed_profiles =
-      feed_profiles_q(user_id, gender, gender_preferences, feed_cursor)
-      |> limit(^count)
-      |> select([p], {p, distance_km(^location, p.location)})
-      |> Repo.all()
+    case feed_cursor do
+      nil -> save_eligible_feed_profiles_to_memory(user_id, location, gender, gender_preferences)
+      _ -> :ok
+    end
+
+    feed_profiles_with_cursor = feed_profiles_from_memory(user_id, feed_cursor, count)
 
     feed_cursor =
-      if last = List.last(feed_profiles) do
-        {feed_profile, _distance} = last
-        feed_profile.last_active
+      if last = List.last(feed_profiles_with_cursor) do
+        {cursor, _feed_profile} = last
+        cursor
       else
         feed_cursor
       end
 
+    feed_profiles =
+      feed_profiles_with_cursor
+      |> Enum.map(fn {_cursor, profile} ->
+        profile
+      end)
+
     {feed_profiles, feed_cursor}
   end
 
+  def save_eligible_feed_profiles_to_memory(user_id, location, gender, gender_preferences) do
+    table = String.to_atom("feed_#{user_id}")
+
+    case :ets.whereis(table) do
+      :undefined -> :ok
+      _ -> :ets.delete(table)
+    end
+
+    :ets.new(table, [:ordered_set, :named_table, read_concurrency: true])
+
+    eligible_feed_profiles =
+      feed_profiles_q(user_id, gender, gender_preferences)
+      |> select([p], {p, distance_km(^location, p.location)})
+      |> Repo.all()
+
+    most_liked_feed_profiles =
+      Enum.sort_by(eligible_feed_profiles, fn {p, _} -> p.times_liked end, :desc)
+
+    recently_active_feed_profiles =
+      Enum.sort_by(eligible_feed_profiles, fn {p, _} -> p.last_active end, {:desc, DateTime})
+
+    feed_profiles = mix_feed_profiles(most_liked_feed_profiles, recently_active_feed_profiles)
+
+    for feed_profile <- feed_profiles do
+      id = Ecto.Bigflake.UUID.generate()
+      :ets.insert(table, {id, feed_profile})
+    end
+  end
+
+  defp mix_feed_profiles(set1, set2) do
+    feed_profiles =
+      Enum.zip(set1, set2)
+      |> Enum.flat_map(fn {p1, p2} -> [p1, p2] end)
+
+    Enum.uniq(feed_profiles)
+  end
+
+  def feed_profiles_from_memory(user_id, feed_cursor, count) do
+    table_name = String.to_atom("feed_#{user_id}")
+
+    keys = next(table_name, feed_cursor, count)
+
+    feed_profiles =
+      keys
+      |> Enum.map(fn key ->
+        :ets.lookup(table_name, key) |> Enum.at(0)
+      end)
```

Comment:
table is needlessly iterated twice, once in `next` and then again for all keys in `lookup`